parquet-common/src/main/java/org/apache/parquet/bytes/SingleBufferInputStream.java

```diff
@@ -0,0 +1,177 @@
+    // duplicate the buffer because its state will be modified
+    this.buffer = buffer.duplicate();
+    this.startPosition = buffer.position();
+  }
+
+  @Override
+  public long position() {
+    // position is relative to the start of the stream, not the buffer
+    return buffer.position() - startPosition;
+  }
+
+  @Override
+  public int read() throws IOException {
+    if (!buffer.hasRemaining()) {
+    	throw new EOFException();
+    }
+    return buffer.get() & 0xFF; // as unsigned
+  }
+
+  @Override
+  public int read(byte[] bytes, int offset, int length) throws IOException {
+    if (length == 0) {
+      return 0;
+    }
+
+    int remaining = buffer.remaining();
+    if (remaining <= 0) {
+      return -1;
+    }
+
+    int bytesToRead = Math.min(buffer.remaining(), length);
+    buffer.get(bytes, offset, bytesToRead);
+
+    return bytesToRead;
+  }
+  
+  @Override
+  public long skip(long n) {
+    if (n == 0) {
+      return 0;
+    }
+
+    if (buffer.remaining() <= 0) {
+      return -1;
+    }
+
+    // buffer.remaining is an int, so this will always fit in an int
+    int bytesToSkip = (int) Math.min(buffer.remaining(), n);
+    buffer.position(buffer.position() + bytesToSkip);
+
+    return bytesToSkip;
+  }
+
+  @Override
+  public int read(ByteBuffer out) {
+    int bytesToCopy;
+    ByteBuffer copyBuffer;
+    if (buffer.remaining() <= out.remaining()) {
+      // copy all of the buffer
+      bytesToCopy = buffer.remaining();
+      copyBuffer = buffer;
+    } else {
+      // copy a slice of the current buffer
+      bytesToCopy = out.remaining();
+      copyBuffer = buffer.duplicate();
+      copyBuffer.limit(buffer.position() + bytesToCopy);
+      buffer.position(buffer.position() + bytesToCopy);
+    }
+
+    out.put(copyBuffer);
+    out.flip();
+
+    return bytesToCopy;
+  }
+
+  @Override
+  public ByteBuffer slice(int length) throws EOFException {
+    if (buffer.remaining() < length) {
+      throw new EOFException();
+    }
+
+    // length is less than remaining, so it must fit in an int
+    ByteBuffer copy = buffer.duplicate();
+    copy.limit(copy.position() + length);
+    buffer.position(buffer.position() + length);
+
+    return copy;
+  }
+
+  @Override
+  public List<ByteBuffer> sliceBuffers(long length) throws EOFException {
+    if (length == 0) {
+      return Collections.emptyList();
+    }
+
+    if (length > buffer.remaining()) {
+      throw new EOFException();
+    }
+
+    // length is less than remaining, so it must fit in an int
+    return Collections.singletonList(slice((int) length));
+  }
+
+  @Override
+  public List<ByteBuffer> remainingBuffers() {
+    if (buffer.remaining() <= 0) {
+      return Collections.emptyList();
+    }
+
+    ByteBuffer remaining = buffer.duplicate();
+    buffer.position(buffer.limit());
+
+    return Collections.singletonList(remaining);
+  }
+
+  @Override
+  public synchronized void mark(int readlimit) {
```

Comment:
This is from `InputStream` and can't be changed.

Reply:
I don't see why it cannot be changed. Whether or whether not the super method is synchronized it is not part of the signature. `InputStream` is implemented to be thread-safe but this implementation is clearly not. I think, using synchronized here suggests that this implementation is also thread-safe which is misleading. In addition it might have some performance penalty as well but I don't think it is measurable.

Reply:
I think you're right and they're not in the method signature. I've removed them and we'll see if the tests pass.